xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/LocalMUCRoom.java

```diff
@@ -2933,35 +2933,56 @@ public void groupModified(Group group, Map params) {
 
     @Override
     public void memberAdded(Group group, Map params) {
-        applyAffiliationChangeAndSendPresence(new JID((String)params.get("member")));
+        applyAffiliationChangeAndSendPresence(new JID((String)params.get("member")), "member");
     }
 
     @Override
     public void memberRemoved(Group group, Map params) {
-        applyAffiliationChangeAndSendPresence(new JID((String)params.get("member")));
+        applyAffiliationChangeAndSendPresence(new JID((String)params.get("member")), "none");
     }
 
     @Override
     public void adminAdded(Group group, Map params) {
-        applyAffiliationChangeAndSendPresence(new JID((String)params.get("admin")));
+        applyAffiliationChangeAndSendPresence(new JID((String)params.get("admin")), "member");
     }
 
     @Override
     public void adminRemoved(Group group, Map params) {
-        applyAffiliationChangeAndSendPresence(new JID((String)params.get("admin")));
+        applyAffiliationChangeAndSendPresence(new JID((String)params.get("admin")), "none");
     }
-    
-    private void applyAffiliationChangeAndSendPresence(JID groupMember) {
+
+    private void applyAffiliationChangeAndSendPresence(JID groupMember, String affiliation) {
         List<Presence> presences = applyAffiliationChange(getRole(), groupMember, null);
-        for (Presence presence : presences) {
-            send(presence);
+
+        if (presences.size() == 0 && isMembersOnly()) {
+            sendAffiliationChangeNotification(groupMember, affiliation);
+        }
+        else {      // member is in MUC, send presence stanzas
+            for (Presence presence : presences) {
+                send(presence);
+            }
         }
     }
 
+    private void sendAffiliationChangeNotification(JID jid, String affiliation) {
+        // Announce affiliation change for a user that is NOT currently in the room,
+        // XEP-0045 (v1.31.2) example 195
+
+        Message message = new Message();
+        Element fragment = message.addChildElement("x", "http://jabber.org/protocol/muc#user");
+        Element item = fragment.addElement("item");
+        item.addAttribute("affiliation", affiliation);
+        item.addAttribute("role", "none");
```

Comment:
I don't think it's correct to add the `role` attribute like this. This does not take into account whatever role the user had - it will always inform others that this user now has no role.

Reply:
I agree, but the user is not in the room. His role will be updated when the join room event happens.

Reply:
>It'd also be good to add javadoc to the methods that you introduced.

I agree, but there is no javadoc on any of the existing methods in that class. I will tackle that later for the entire class.

Reply:
>I think. It might be good to strictly limit the commit to non-white-line changes (maybe add the white-line changes to a different commit?)

This is my third attempt at this commit. My editor was removing redundant white spaces introduced by a previous commit. I am not sure I have time to start a battle with my text editor.

Reply:
Tip: move all the white-space changes to the first commit, so that everything you touch afterwards does not need to change whitespace!

Reply:
> I agree, but the user is not in the room. His role will be updated when the join room event happens.

Even though that's not how you call the method now, there's nothing that prevents future developers from calling it when the person _is_ in the room.

Reply:
Got it. I will rename the method accordingly